### DISCUSSION TITLE
goog.reflect.sinkValue and deadcode elimination

### DISCUSSION THREAD

**Discussion by treblereel** [google/j2cl#203] [Q&A] [answered]
One of the tips for preventing code removal by the compiler is the use of reflect.sinkValue from closure-library. Is it possible to use this feature without resorting to creating native.js?

lets say I have something like this:
```
public class Test {
    Random random = new Random();
    private void test() {
        DomGlobal.console.log("test "  + random.nextInt(10) + this.getClass().getCanonicalName());
    }
}
```
And I want to call it like this:
```
        Test test = new Test();
        ((Function)Js.asPropertyMap(test).get(objectProperty("m_test___$p_org_treblereel_Test", test))).bind(test).call();
         // it works for non inlined/removed functions       
```
```
    @JsMethod(namespace = "goog.reflect")
    public native static String objectProperty(String property, Object object);
```

What should I pass to sinkValue wrapper?
```
    @JsMethod(namespace = "goog.reflect")
    public native static Object sinkValue(??? value);
```
As I can see from the docs I should use something like ```sinkValue(test.m_test___$p_org_treblereel_Test)``` but it's not possible from java.

**Comment by gkdn** [comment, marked as answer]
There are couple of issues here.

The reason you are looking into sinkValue is because you are not directly accessing the property, but accessing it through a map which is a string access. Your code is effectively `test["m_test___$p_org_treblereel_Test"]` which wouldn't be recognized as a property access by JsCompiler hence will be renamed, removed etc. If the access looked like `test.test` then you wouldn't need goog.reflect.objectProperty or sinkValue. [[Details]](https://developers.google.com/closure/compiler/docs/api-tutorial3#propnames)

The second issues here is that the mangled names like `m_test___$p_org_treblereel_Test` are not stable. We are not intending to preserve the mangling rules and we already have changed them in the past. So this may get broken any time. To control and have predictable names you really need to use JsInterop annotations (`JsType` `JsMethod` etc.).

You are trying work around dead code elimination but if you can describe what you would like to really achieve, perhaps there are better ways to do what you want.

However for the sake of completeness and answering your question directly, following should work with sinkValue :
```
@JsType(isNative=true, namespace="mypackage", name="Test")
class TestProxy {
  @JsProperty(name="m_test___$p_org_treblereel_Test")
  Function testProperty;
}

sinkValue(((TestProxy) test).testProperty);
```

If you are doing this, then you can as well just `bind` without sinkValue since property access will be recognized and won't be removed:
```
((TestProxy) test).testProperty.bind(test).call();
```

If you can modify Test class you can also do:
```
public class Test {
   @JsMethod
   private void test() {
     // ...
   }
   @JsProperty
   native Function getTest();
}

test.getTest().bind(test).call();
```

**Comment by treblereel** [comment]
@gkdn thank you, works great!
I am working on j2cl replacement of redhat's errai framework, so I need to be able to process beans like this one:

```
@ApplicationScoped
public class Test {
    Random random = new Random();
    @PostConstuct
    private void test() {
        DomGlobal.console.log("test "  + random.nextInt(10) + this.getClass().getCanonicalName());
    }
}
```
To generate a mangled name I reuse com.google.j2cl.transpiler.ast.MemberDescriptor, so I have to be in sync with j2cl version. 
It's not the best option, but like I have a choice :)
